### PR TITLE
Clickhouse support all join types 

### LINKED ISSUE #3506: ClickHouse dialect does not support all join types
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

ClickHouse supports more join types than ANSI SQL, including `semi join` and `anti join`. Their docs show a join clause syntax like
```
SELECT <expr_list>
FROM <left_table>
[GLOBAL] [INNER|LEFT|RIGHT|FULL|CROSS] [OUTER|SEMI|ANTI|ANY|ASOF] JOIN <right_table>
(ON <expr_list>)|(USING <column_list>) ...
```
([Source](https://clickhouse.com/docs/en/sql-reference/statements/select/join/#supported-types-of-join))

However, SQLFluff misinterprets the `semi` in `semi join`, for example, as a table alias. It doesn't seem to support join types beyond typical ones like `left join` or `inner join`.

### Expected Behaviour

The tool should understand all of the join types in the syntax above.

### Observed Behaviour

A join query like this one:

```sql
select
    example1.a,
    b
from example1
    semi join example2 on (example1.a = example2.a);
```

yields these warnings (note how it thinks `semi` is an alias for `example1`, while it's really a join type like `left join` or `inner join`):

```
❯ sqlfluff lint example.sql --dialect clickhouse
== [example.sql] FAIL
L:  19 | P:   5 | L027 | Unqualified reference 'b' found in select with more than
                       | one referenced table/view.
L:  21 | P:   5 | L011 | Implicit/explicit aliasing of table.
L:  21 | P:   5 | L025 | Alias 'semi' is never used in SELECT statement.
L:  21 | P:   5 | L031 | Avoid aliases in from clauses and join conditions.
L:  21 | P:  10 | L051 | Join clauses should be fully qualified.
All Finished 📜 🎉!
```

### How to reproduce

```sql
-- example.sql
select
    example1.a,
    b
from example1
    semi join example2 on (example1.a = example2.a);
```

Command:

```
sqlfluff lint example.sql --dialect clickhouse
```

### Dialect

ClickHouse

### Version

```
❯ sqlfluff --version
sqlfluff, version 1.0.0

❯ python --version
Python 3.7.13
```

### Configuration

Just the stock configuration, no other config files.

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by konnectr**
### Brief summary of the change made

Closes #3506

### Are there any other side effects of this change that we should be aware of?

The reserved names in Clickhouse will match the reserved names in ansi.

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

**Comment by konnectr** [comment]
I have a question, how best to cover this case.
At Clickhouse:
`SELECT * FROM test1 LEFT ANTI JOIN test2 USING ty1,ty2;`
equal
`SELECT * FROM test1 ANTI LEFT  JOIN test2 USING ty1,ty2;`

I have made a dialect for the moment only for the first option. What is the best way to support two cases?

File: src/sqlfluff/dialects/dialect_clickhouse.py [hunk 1]
```diff
@@ -25,7 +28,96 @@
...
+                Sequence(
+                    Ref.keyword("RIGHT"),
+                    OneOf(
+                        "OUTER",
+                        "ANTI",
+                        "SEMI",
+                        "ANY",
+                        optional=True,
+                    ),
+                ),
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_clickhouse.py:65]:
```suggestion
                Sequence(
                    "LEFT",
                    OneOf(
                        "OUTER",
                        "ANTI",
                        "SEMI",
                        "ANY",
                        "ASOF",
                        optional=True,
                    ),
                ),
                Sequence(
                    "ANTI",
                    "LEFT",
                ),
                Sequence(
                    "RIGHT",
                    OneOf(
                        "OUTER",
                        "ANTI",
                        "SEMI",
                        "ANY",
                        optional=True,
                    ),
                ),
                Sequence(
                    "ANTI",
                    "RIGHT",
                ),
```

File: src/sqlfluff/dialects/dialect_clickhouse.py [hunk 2]
```diff
@@ -25,7 +28,96 @@
...
+
+    match_grammar = OneOf(
+        Sequence(
+            Ref("JoinTypeKeywords", optional=True),
+            Ref("JoinKeywordsGrammar"),
+            Indent,
+            Ref("FromExpressionElementSegment"),
+            Dedent,
+            Conditional(Indent, indented_using_on=True),
+            # NB: this is optional
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_clickhouse.py:88]: Hence why we have optional attribute so don't think you need this.

```suggestion
```

**Review by tunetheweb — CHANGES_REQUESTED** [review]
LGTM but a couple of minor suggestions

**Comment by tunetheweb** [comment]
> I have a question, how best to cover this case. At Clickhouse: `SELECT * FROM test1 LEFT ANTI JOIN test2 USING ty1,ty2;` equal `SELECT * FROM test1 ANTI LEFT JOIN test2 USING ty1,ty2;`
> 
> I have made a dialect for the moment only for the first option. What is the best way to support two cases? @tunetheweb

Made a suggestion in the code review as to how it could be covered. I don't know if only `ANTI` is supported like this, or if the full pre join keyword should have been repeated?

File: src/sqlfluff/dialects/dialect_clickhouse.py (hunk 1 above)
Comment (konnectr) [reply src/sqlfluff/dialects/dialect_clickhouse.py]: I changed the Sequence

File: src/sqlfluff/dialects/dialect_clickhouse.py [hunk 3]
```diff
@@ -71,6 +213,7 @@ class FromExpressionElementSegment(ansi.FromExpressionElementSegment):
                 Ref("SamplingExpressionSegment"),
                 Ref("JoinLikeClauseGrammar"),
                 Ref.keyword("Final"),
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_clickhouse.py:215]: Not your issue but just noticed this. IT doesn't need the Ref.keyword and `Final` should be in uppercase.

```suggestion
                FINAL,
```

**Review by tunetheweb — APPROVED** [review]

File: src/sqlfluff/dialects/dialect_clickhouse.py [hunk 4]
```diff
@@ -70,7 +212,8 @@ class FromExpressionElementSegment(ansi.FromExpressionElementSegment):
             exclude=OneOf(
                 Ref("SamplingExpressionSegment"),
                 Ref("JoinLikeClauseGrammar"),
-                Ref.keyword("Final"),
+                FINAL,
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_clickhouse.py:215]:
```suggestion
                "FINAL",
```

**Comment by tunetheweb** [comment]
@konnectr it looks like ST07 has an issue with `USING` clauses that do not uses brackets - like some of your queries here. This was causing one of the CI jobs to fail. I've therefore disabled this rule for the Clickhouse dialect in https://github.com/sqlfluff/sqlfluff/pull/4488/commits/ef40cc90919a2110a00bccec404fe731a82cab33

Hoping CI passes now!

**Pull request merged**